dali/operators/random/uniform_distribution.h

```diff
@@ -0,0 +1,173 @@
+      // read only once for build time arguments
+      if (!values_.IsConstant() || per_sample_values_.empty()) {
+        per_sample_values_.resize(nsamples);
+        per_sample_nvalues_.resize(nsamples);
+        if (std::is_same<Backend, GPUBackend>::value) {
+          values_cpu_.clear();
+          for (int s = 0; s < nsamples; s++) {
+            values_cpu_.insert(values_cpu_.end(),
+                               values_[s].data,
+                               values_[s].data + values_[s].shape[0]);
+            per_sample_nvalues_[s] = values_[s].shape[0];
+          }
+          values_gpu_.from_host(values_cpu_, ws.stream());
+          int64_t offset = 0;
+          for (int s = 0; s < nsamples; s++) {
+            per_sample_values_[s] = values_gpu_.data() + offset;
+            offset += per_sample_nvalues_[s];
+          }
+        } else {
+          for (int s = 0; s < nsamples; s++) {
+            per_sample_values_[s] = values_[s].data;
+            per_sample_nvalues_[s] = values_[s].shape[0];
+          }
+        }
+      }
+    } else {
+      range_.Acquire(spec, ws, nsamples, TensorShape<1>{2});
+    }
+  }
+
+  DALIDataType DefaultDataType() const {
+    return DALI_FLOAT;
+  }
+
+  template <typename T>
+  bool SetupDists(typename DistContinuous<T>::type* dists, int nsamples) {
+    for (int s = 0; s < nsamples; s++) {
+      dists[s] = typename DistContinuous<T>::type(range_[s].data[0], range_[s].data[1]);
+    }
+    // note: can't use the default because this operator's default range is different to
+    // the default constructed distribution.
+    return true;
+  }
+
+  template <typename T>
+  bool SetupDists(typename DistDiscrete<T>::type* dists, int nsamples) {
+    assert(values_.IsDefined());
+    for (int s = 0; s < nsamples; s++) {
+      dists[s] = typename DistDiscrete<T>::type(per_sample_values_[s], per_sample_nvalues_[s]);
+    }
+    return true;
+  }
+
+  template <typename T>
+  void RunImplTyped(workspace_t<Backend> &ws) {
+    using Base = RNGBase<Backend, UniformDistribution<Backend>>;
+    if (values_.IsDefined()) {
+      using Dist = typename DistDiscrete<T>::type;
+      Base::template RunImplTyped<T, Dist>(ws);
+    } else {
+      using Dist = typename DistContinuous<T>::type;
+      Base::template RunImplTyped<T, Dist>(ws);
+    }
+  }
+
+  void RunImpl(workspace_t<Backend> &ws) override {
+    TYPE_SWITCH(dtype_, type2id, T, DALI_UNIFORM_DIST_TYPES, (
+      this->template RunImplTyped<T>(ws);
+    ), DALI_FAIL(make_string("Unsupported data type: ", dtype_)));  // NOLINT
+  }
+
+ protected:
+  using Operator<Backend>::max_batch_size_;
+  using RNGBase<Backend, UniformDistribution<Backend>>::dtype_;
+  using RNGBase<Backend, UniformDistribution<Backend>>::backend_data_;
+
+  ArgValue<float, 1> values_;
+  ArgValue<float, 1> range_;
+
+  std::vector<float> values_cpu_;
+  DeviceBuffer<float> values_gpu_;
+  std::vector<const float*> per_sample_values_;
```

Comment:
Do you free the `const float*` arrays properly?

Reply:
Those are just per sample pointers to a position in either values_gpu_ or values_cpu_